controller/volume_controller.go

```diff
@@ -1125,77 +1182,401 @@ func (vc *VolumeController) ReconcileVolumeState(v *longhorn.Volume, es map[stri
+					// so we just unset PendingNodeID.
+					//v.Status.PendingNodeID = v.Spec.NodeID
+					v.Status.RemountRequestedAt = vc.nowHandler()
+					msg := fmt.Sprintf("Volume %v requested remount at %v after automatically salvaging replicas", v.Name, v.Status.RemountRequestedAt)
+					vc.eventRecorder.Eventf(v, v1.EventTypeNormal, constant.EventReasonRemount, msg)
+					v.Status.Robustness = longhorn.VolumeRobustnessUnknown
+					return nil
+				}
+			}
+		}
+	} else { // !isAutoSalvageNeeded
+		if v.Status.Robustness == longhorn.VolumeRobustnessFaulted && v.Status.State == longhorn.VolumeStateDetached {
+			v.Status.Robustness = longhorn.VolumeRobustnessUnknown
+			// The volume was faulty and there are usable replicas.
+			// Therefore, we set RemountRequestedAt so that KubernetesPodController restarts the workload pod
+			v.Status.RemountRequestedAt = vc.nowHandler()
+			msg := fmt.Sprintf("Volume %v requested remount at %v", v.Name, v.Status.RemountRequestedAt)
+			vc.eventRecorder.Eventf(v, v1.EventTypeNormal, constant.EventReasonRemount, msg)
+			return nil
 		}
 
-		return fmt.Errorf("volume %v has already attached to node %v, but asked to attach to node %v", v.Name, v.Status.CurrentNodeID, v.Spec.NodeID)
+		// reattach volume if detached unexpected and there are still healthy replicas
+		if e.Status.CurrentState == longhorn.InstanceStateError && v.Status.CurrentNodeID != "" {
+			log.Warn("Engine of volume dead unexpectedly, reattach the volume")
+			msg := fmt.Sprintf("Engine of volume %v dead unexpectedly, reattach the volume", v.Name)
+			vc.eventRecorder.Event(v, v1.EventTypeWarning, constant.EventReasonDetachedUnexpectly, msg)
+			e.Spec.LogRequested = true
+			for _, r := range rs {
+				if r.Status.CurrentState == longhorn.InstanceStateRunning {
+					r.Spec.LogRequested = true
+					rs[r.Name] = r
+				}
+			}
+			v.Status.Robustness = longhorn.VolumeRobustnessFaulted
+		}
 	}
 
-	if e == nil {
+	if err := vc.reconcileAttachDetachStateMachine(v, e, rs, isNewVolume, log); err != nil {
+		return err
+	}
+
+	if v.Status.CurrentNodeID != "" &&
+		v.Status.State == longhorn.VolumeStateAttached &&
+		e.Status.CurrentState == longhorn.InstanceStateRunning {
+		if e.Spec.RequestedBackupRestore != "" {
+			v.Status.Conditions = types.SetCondition(v.Status.Conditions,
+				longhorn.VolumeConditionTypeRestore, longhorn.ConditionStatusTrue, longhorn.VolumeConditionReasonRestoreInProgress, "")
+		}
+
+		// TODO: reconcileVolumeSize
+		// The engine expansion is complete
+		if v.Status.ExpansionRequired && v.Spec.Size == e.Status.CurrentSize {
+			v.Status.ExpansionRequired = false
+			v.Status.FrontendDisabled = false
+		}
+	}
+
+	if err := vc.checkAndFinishVolumeRestore(v, e, rs); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (vc *VolumeController) reconcileAttachDetachStateMachine(v *longhorn.Volume, e *longhorn.Engine, rs map[string]*longhorn.Replica, isNewVolume bool, log *logrus.Entry) error {
```

Comment:
The state machine transition logics are understandable but a bit lengthy.
Suggest abstracting the blocks to different functions with meaningful names and make them more concise.

Reply:
Do you mean moving the switch statements into different functions? Could you give an example? Thank you very much